lib/elixir/lib/module/types/infer.ex

```diff
@@ -102,26 +104,84 @@ defmodule Module.Types.Infer do
 
   # %{...}
   def of_pattern({:%{}, _meta, args} = expr, context) do
-    result =
-      expr_stack(expr, context, fn context ->
-        map_reduce_ok(args, context, fn {key, value}, context ->
-          with {:ok, key_type, context} <- of_pattern(key, context),
-               {:ok, value_type, context} <- of_pattern(value, context),
-               do: {:ok, {key_type, value_type}, context}
-        end)
+    case expr_stack(expr, context, &of_pairs(args, &1)) do
+      {:ok, pairs, context} -> {:ok, {:map, pairs_to_unions(pairs, context)}, context}
+      {:error, reason} -> {:error, reason}
+    end
+  end
+
+  # %var{...}
+  def of_pattern({:%, _meta1, [var, {:%{}, _meta2, args}]} = expr, context) when is_var(var) do
+    expr_stack(expr, context, fn context ->
+      with {:ok, pairs, context} <- of_pairs(args, context),
+           {var_type, context} = new_var(var, context),
+           {:ok, _, context} <- unify(var_type, :atom, context) do
+        pairs = [{{:literal, :__struct__}, var_type} | pairs]
+        {:ok, {:map, pairs_to_unions(pairs, context)}, context}
+      end
+    end)
+  end
+
+  # %Struct{...}
+  def of_pattern({:%, _meta1, [module, {:%{}, _meta2, args}]} = expr, context)
+      when is_atom(module) do
+    struct_pairs =
+      Enum.map(:maps.remove(:__struct__, module.__struct__()), fn {key, value} ->
+        {term_to_type(key), term_to_type(value)}
       end)
 
-    case result do
-      {:ok, pairs, context} -> {:ok, {:map, pairs}, context}
-      {:error, reason} -> {:error, reason}
+    case expr_stack(expr, context, &of_pairs(args, &1)) do
+      {:ok, pattern_pairs, context} ->
+        pairs = pairs_to_unions(pattern_pairs ++ struct_pairs, context)
+        {:ok, {:map, pairs}, context}
```

Comment:
It would probably be best to keep it as `{:strut, name, pairs}`. Knowing if something is a struct or not will definitely be handy. Not sure how we should represent `%var{}` though...

Reply:
`%var{}` would be represented as `{:struct, {:var, 0}, pairs}`.

Reply:
The issue is that we don't know anything about the struct, the fields it has, what would be their types... so we won't be able to check with the same level of precision as regular ones. But I guess it is best to not discard any information early on... and do so later only if necessary.